File: problems/Next Permutation /nextPermutation.cs
public class Solution {
    public void NextPermutation(int[] nums) {
        var n = nums.Length - 1;
        var i = n - 1;

        while (0 <= i && nums[1 + i] <= nums[i]) {--i;}

        if (0 <= i) {
            var j = n;

            while (i < j && nums[i] >= nums[j]) {--j;}

            Swap(nums, i, j);
        }

        Reverse(nums, 1 + i, n);
    }

    private void Swap(int[] arr, int i, int j) {
        var temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private void Reverse(int[] arr, int i, int j) {
        while (i < j) {
            Swap(arr, i++, j--);
        }
    }
}

File: problems/Next Permutation /nextPermutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int left = n - 2;
        // [3,5,4,2,1] -> [4,1,2,3,5]
        while (0 <= left && nums[left] >= nums[1 + left]) {
            --left;
        }
        // left: 0
        if (-1 != left) {
            int right = n - 1;

            while (0 <= right && nums[left] >= nums[right]) {
                --right;
            }
            // right: 2
            swap(nums, left, right);
            // [4,5,3,2,1]
        }

        reverse(nums, 1 + left, n - 1);
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    private void reverse(int[] nums, int i, int j) {
        while (i < j) {
            swap(nums, i, j);
            ++i;
            --j;
        }
    }
}
